### PR TITLE
Propagate FCU invalid error

### PR DESCRIPTION
**Pull request by terencechain**
In the event, FCU returns INVALID. We should propagate FCU invalid error back to the callers so the callers could mark the block as bad and blacklist for future children
I also wrapped `invalidBlock` on the original error definitions rather than on the return. Less disparity

### CODE REVIEW

File: beacon-chain/blockchain/execution_engine.go [hunk 1]
```diff
@@ -212,10 +212,10 @@ func (s *Service) notifyNewPayload(ctx context.Context, postStateVersion int,
 			"blockRoot":    fmt.Sprintf("%#x", root),
 			"invalidCount": len(invalidRoots),
 		}).Warn("Pruned invalid blocks")
-		return false, invalidBlock{ErrInvalidPayload}
+		return false, ErrInvalidPayload
```
Comment (terencechain) [inline beacon-chain/blockchain/execution_engine.go:215]: Already wrapped with `invalidBlock`. See beacon-chain/blockchain/error.go

File: beacon-chain/blockchain/receive_attestation.go [hunk 2]
```diff
@@ -202,10 +204,14 @@ func (s *Service) notifyEngineIfChangedHead(ctx context.Context, newHeadRoot [32
 	_, err = s.notifyForkchoiceUpdate(s.ctx, arg)
 	if err != nil {
 		log.WithError(err).Error("could not notify forkchoice update")
+		if IsInvalidBlock(err) {
```
Comment (terencechain) [inline beacon-chain/blockchain/receive_attestation.go:207]: To not alter with original behavior too much, only return error if the block has invalid payload for blacklist

Comment (potuz) [reply beacon-chain/blockchain/receive_attestation.go:207]: This will log twice the same error, here and then in the outer caller when `err` surfaces. Perhaps consider putting this block inside an if-else block.

File: beacon-chain/blockchain/receive_attestation.go [hunk 3]
```diff
@@ -202,10 +204,14 @@ func (s *Service) notifyEngineIfChangedHead(ctx context.Context, newHeadRoot [32
...
+			return err
```
Comment (potuz) [inline beacon-chain/blockchain/receive_attestation.go:208]: I feel like either we need to propagate the error here, or log and stop propagation within `notifyForkchoiceUpdate`, doing it middle way seems like it will lead to bugs. What would be the problem in propagating here? or what would be the problem on only returning errors from `notifyForkchoiceUpdate` when `IsInvalidBlock(err) == true`?

File: beacon-chain/blockchain/execution_engine.go [hunk 4]
```diff
@@ -77,37 +81,46 @@ func (s *Service) notifyForkchoiceUpdate(ctx context.Context, arg *notifyForkcho
...
 			invalidRoots, err := s.ForkChoicer().SetOptimisticToInvalid(ctx, headRoot, bytesutil.ToBytes32(headBlk.ParentRoot()), bytesutil.ToBytes32(lastValidHash))
 			if err != nil {
-				return nil, err
+				log.WithError(err).Error("Could not set head root to invalid")
+				return nil, nil
 			}
 			if err := s.removeInvalidBlockAndState(ctx, invalidRoots); err != nil {
-				return nil, err
+				log.WithError(err).Error("Could not remove invalid block and state")
+				return nil, nil
```
Comment (potuz) [inline beacon-chain/blockchain/execution_engine.go:99]: same here

File: beacon-chain/blockchain/execution_engine.go [hunk 5]
```diff
@@ -77,37 +81,46 @@ func (s *Service) notifyForkchoiceUpdate(ctx context.Context, arg *notifyForkcho
...
 			}
 
 			r, err := s.cfg.ForkChoiceStore.Head(ctx, s.justifiedBalances.balances)
 			if err != nil {
-				return nil, err
+				log.WithError(err).Error("Could not get head root")
+				return nil, nil
```
Comment (potuz) [inline beacon-chain/blockchain/execution_engine.go:105]: and here

File: beacon-chain/blockchain/execution_engine.go [hunk 6]
```diff
@@ -77,37 +81,46 @@ func (s *Service) notifyForkchoiceUpdate(ctx context.Context, arg *notifyForkcho
...
 			}
 			b, err := s.getBlock(ctx, r)
 			if err != nil {
-				return nil, err
+				log.WithError(err).Error("Could not get head block")
+				return nil, nil
```
Comment (potuz) [inline beacon-chain/blockchain/execution_engine.go:110]: and here

File: beacon-chain/blockchain/error.go [hunk 7]
```diff
@@ -41,18 +41,25 @@ var (
...
+	BlockRoot() [32]byte
 }
 
 // InvalidBlock returns true for `invalidBlock`.
 func (e invalidBlock) InvalidBlock() bool {
 	return true
 }
 
+// BlockRoot returns the invalid block root.
+func (e invalidBlock) BlockRoot() [32]byte {
```
Comment (potuz) [inline beacon-chain/blockchain/error.go:59]: Why is this needed instead of just returning `true` in the last line of `IsInvalidBlock`? (like 72)

**Review by potuz — APPROVED** [review]
